coccigrep.1
.TH coccigrep 1 "October 16, 2011" "version 1.3"
.SH NAME
coccigrep - semantic grep for the C programming language

.SH SYNOPSIS
.B coccigrep
.RI [ OPTIONS ]
.BI "\-t " "TYPE "
.RI [ FILE .\|.\|.]

.SH DESCRIPTION
.BR coccigrep " is a semantic " grep " for the C programming language."
.RB "It is based on " coccinelle " and can be used to find where a given"
.RB "structure is used in C source files. " coccigrep " depends"
.RB "on the " spatch " program which comes with coccinelle."

.SH OPTIONS
.SS Display options
.TP
.BI \-A " NUM" "\fR,\fP \-\^\-after\-context=" NUM
.RI "Print " NUM " lines of trailing context after matching line."
.TP
.BI \-B " NUM" "\fR,\fP \-\^\-before\-context=" NUM
.RI "Print " NUM " lines of leading context before matching line."
.TP
.BR \-c ", " \-\-color
Colorize output (needs pygments).
.TP
.BI \-C " NUM" "\fR,\fP \-\^\-context=" NUM
.RI "Print " NUM " lines of output context."
.TP
.BI \-f " FORMAT" "\fR,\fP \-\^\-output-format=" FORMAT
Colorize format for output.
.I FORMAT
.RB "is " term " or " html .

.SS Search options
.TP
.BI \-a " ATTRIBUTE" "\fR,\fP \-\^\-attribut=" ATTRIBUTE
.RI "Match operations that involve the field " ATTRIBUTE " of the"
.RB "structure given with option " \-t "."
.TP
.B \-\-cpp
Activate coccinelle C++ support.
.TP
.BI \-o " OPERATION" "\fR,\fP \-\^\-operation=" OPERATION
Specify the kind of operation to look for.
.RI "The available values for " OPERATION " are:"

.RS
.B deref
.RS
.RI "Any access to an " ATTRIBUTE .
.RE
.B func
.RS
.RI "Any function call for which one of the parameters is of type " TYPE
.RB "(as defined with " \-t "). In this operation, the option " \-a " sets"
the name of the function to match and does not refer to an attribute.
.RE
.B set
.RS
.RI "Any operation wich changes the value of an " ATTRIBUTE .
.RE
.B test
.RS
.RI "Any boolean test where an " ATTRIBUTE " appears."
.RE
.B used
.RS
.RI "Any statement where the type " TYPE " (as defined with"
.BR \-t ") appears."
.RI "This " OPERATION " does not need to look for an " ATTRIBUTE " thus"
.RB "it ignores option " \-a .
.RE

.RB "The default operation is " used " if " \-a " is not given,"
.RB "or " deref " if it is."
.RI "Depending on your environment, more " OPERATION "s may be available."
.RB "See option " \-L " for information about"
.IR OPERATION " listing."
.RE
.TP
.BI \-t " TYPE" "\fR,\fP \-\^\-type=" TYPE
C type to look for.

.SS Editor options
.TP
.BR \-E ", " \-\-emacs
Emacs output.
.TP
.BR \-V ", " \-\-vim
Vim output.

.SS Program information
.TP
.BR \-h ", " \-\-help
Print usage message and exit.
.TP
.BR \-L ", " \-\-list-operations
.RI "List available " OPERATION "s. Use together with"
.BR \-v " to get detailed information about listed"
.IR OPERATION "s."
.TP
.BR \-v ", " \-\-verbose
Verbose output (including coccinelle errors).
.TP
.B \-\-version
.RB "Print the version number of " coccigrep "."

.SS Other options
.TP
.BI \-p " NCPUS" "\fR,\fP \-\^\-process=" NCPUS
Number of cpus to use.
.TP
.BI \-s " SP" "\fR,\fP \-\^\-sp=" SP
Semantic patch to use.
.TP
.BI \-l " FILE" "\fR,\fP \-\^\-file\-list=" FILE
File containing a list of input files.

.SH BUGS
The operation option could lead to some missed match because the semantic
patches used internally may need some improvements.
.RB "Use options " \-L " together with " \-v " to get detailed information"
about available operations.

.RB "Please report issues to <" https://github.com/regit/coccigrep/issues ">."
All ideas are welcome.

.SH EXAMPLES
.nf
.RB "1. Find where in a set of files the structure " Packet " is used:"
	$ coccigrep -t Packet *c

.nf
.RB "2. Finds where in a set of files the " datalink " attribute is used in the structure " Packet ":"
	$ coccigrep -t Packet -a datalink *c

.nf
.RB "3. Finds where the " datalink " attribute of " Packet " is set:"
	$ coccigrep -t Packet -a datalink -o set source*c

.SH FILES
.IR ~/.coccigrep "\tConfiguration file for coccigrep."

.SH SEE ALSO
spatch(1)

.RB "Project homepage: <" http://home.regit.org/software/coccigrep/ ">"


ChangeLog
1.5 (2011/11/03)
 * Allow specyfing a list of input files via a file (-l / --file-list) (Thomas Graf)
 * Fix unmatched patterns when using option "--operation=set" (Guillaume Nault)

1.4 (2011/10/27)
 * Detect attribute setting with unary operator. (Eric Leblond)
 * Manpage update: describe option "-o, --operation" (Guillaume Nault)
 * Improve manpage (Guillaume Nault)

1.3 (2011/10/06)
 * Add manpage for coccigrep. (Guillaume Nault)

1.2 (2011/09/15)
 * unprint cmd (Phlogistique)

1.1 (2011/09/14)
 * Add C++ support. (Eric Leblond)
 * Add 'Revision' info to cocci patches. (Eric Leblond)
 * Add 'Arguments' list in cocci patches. (Eric Leblond)
 * Add 'Arguments' as a keyword for cocci comment. (Eric Leblond)
 * Add information about -L -v in help message. (Eric Leblond)
 * Fix trivial typo. (Eric Leblond)
 * Use python from env. (Eric Leblond)
 * '-L -v' now display a list of operations description. (Eric Leblond)
 * Prepare print of operations informations. (Eric Leblond)
 * Add class to get information from cocci patch. (Eric Leblond)

1.0 (2011/09/12)
 * Add description of operation in comment lines. (Eric Leblond)

1.0rc2 (2011/09/06)
 * Document vim search of 'struct myst' (Eric Leblond)
 * Fix usage of structure without typedef. (Eric Leblond)
 * '-C' option can now be used to set context. (Eric Leblond)
 * Improve display for pointer/non pointer match (Eric Leblond)
 * Add non pointer support to deref. (Eric Leblond)
 * Rename '-C' option to '-p' (Eric Leblond)
 * big cosmit. please test for your sake (Feth AREZKI)
 * try and simplify config reading code (Feth AREZKI)
 * Set version to 1.0rc2. (Eric Leblond)
 * Handle static structure. (Eric Leblond)
 * Fix local cocci dir. (Eric Leblond)
 * Use setuptools in setup.py. (Eric Leblond)

1.0rc1 (2011/09/02)
 * Set version to 1.0rc1 (Eric Leblond)
 * print debug message to stderr. (Eric Leblond)
 * Update MANIFEST.in to include doc. (Eric Leblond)
 * Improve doc file. (Eric Leblond)
 * Improve documentation files. (Eric Leblond)
 * Add docstrings to module. (Eric Leblond)
 * Add some docstrings. (Eric Leblond)
 * Suppress false space. (Eric Leblond)
 * Prepare doc generation via sphinx. (Eric Leblond)
 * Add some docstrings. (Eric Leblond)
 * Handle a serie of failures via exception. (Eric Leblond)
 * put more code in try except to intercept all exceptions (Eric Leblond)
 * Work exception handling (Eric Leblond)
 * Fix typo in split files function. (Eric Leblond)
 * Update setup.py for upload to python repo. (Eric Leblond)

0.9 (2011/08/29)
 * Improve vim usage instruction. (Eric Leblond)
 * Display message about waiting time. (Eric Leblond)
 * vim plugin: Use script variable. (Eric Leblond)
 * vim plugin display warning message when there is no match. (Eric Leblond)
 * cocci-grep.vim is now a vim plugin. (Eric Leblond)
 * Exit with 1 if no match. (Eric Leblond)
 * Add vim module to package. (Eric Leblond)
 * Add operation list to interactive vim function (Eric Leblond)
 * Add option to list available operations. (Eric Leblond)
 * Add documentation of vim module. (Eric Leblond)
 * Vim function for coccigrep (Eric Leblond)
 * Minor fix in emacs module (Eric Leblond)

0.8 (2011/08/27)
 * Add information about configuration file (Eric Leblond)
 * Add explanation about emacs. (Eric Leblond)
 * Add basic cocci-grep emacs module (Eric Leblond)
 * Add Emacs mode. (Eric Leblond)
 * Add option for spatch command (Eric Leblond)
 * Add sanity check to add_operations (Eric Leblond)
 * Add support for local cocci directory (Eric Leblond)
 * Display only one line in vim mode. (Eric Leblond)
 * Config file support (Eric Leblond)
 * Rename badly named variable. (Eric Leblond)
 * Disable concurrency for python without multiprocessing support (Eric Leblond)
 * Introduce -C flag to set concurrency level (Eric Leblond)
 * Introduce concurrency level (Eric Leblond)
 * Suppress empty line at end of output (Eric Leblond)
 * Semantic patchfile can be given on command line. (Eric Leblond)
 * Prepare possibility to add operations from command line (Eric Leblond)

0.7 (2011/08/19)
 * Add check for structure usage in function. (Eric Leblond)
 * List of operations is dynamic (Eric Leblond)
 * add function to retrieve data directory (Eric Leblond)

0.6 (2011/08/18)
 * setup: sdist command is now working (Eric Leblond)
 * Put cocci SMPL into separate files (Eric Leblond)
 * Fix implicit usage of 'deref' instead of 'used' (Eric Leblond)
 * Add indication about vim integration. (Eric Leblond)

0.5 (2011/08/14)
 * Don't use exit in module. (Eric Leblond)
 * Move global variable to class variable. (Eric Leblond)
 * Update python build system. (Eric Leblond)
 * Global refactoring (Eric Leblond)


src/coccigrep.py
# Copyright (C) 2011 Eric Leblond <eric@regit.org>
#
# You can copy, redistribute or modify this Program under the terms of
# the GNU General Public License version 3 as published by the Free
# Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# version 3 along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.

from ConfigParser import SafeConfigParser
from os import unlink, path, listdir, getcwd
from string import Template
from subprocess import Popen, PIPE
from sys import stderr
from tempfile import NamedTemporaryFile
import errno
import re

have_multiprocessing = True
try:
    from multiprocessing import Process, Pipe
except ImportError:
    have_multiprocessing = False


have_pygments = True
try:
    from pygments import highlight
    from pygments.lexers import CLexer
    from pygments.filters import NameHighlightFilter
    from pygments.formatters import Terminal256Formatter, HtmlFormatter
except ImportError:
    have_pygments = False


class CocciException(Exception):
    """
    Generic class for coccigrep exception
    """
    def __init__(self, value):
        self.value = value

    def __str__(self):
        return self.value


class CocciConfigException(CocciException):
    """
    Exception raised when configuration parameter are not correct.

    For example, it is returned if spatch command can not be found.
    """
    pass


class CocciRunException(CocciException):
    """
    Exception raised when running parameters are not correct.

    For example, it is returned if a required argument is missing.
    """
    pass


# produces conf file paths in reverse specificity order, ie
# - package conf
# - system wide
# - user
# - current dir
_CONF_FILES = tuple((path.join(dirname, name_format % 'coccigrep')
    for dirname, name_format in
    ((path.dirname(__file__), '%s.cfg'),
    ('/etc', '%s'),
    (path.expanduser('~'), '.%s'),
    (getcwd(), '.%s'))))


class CocciGrepConfig(SafeConfigParser):
    """
    Configuration handling class

    This class parses configuration and can be used to access to
    configuration item via get operations. CocciGrepConfig is derived
    from SafeConfigParser
    """
    def __init__(self):
        SafeConfigParser.__init__(self)
        self._load_config()

    def _load_config(self):
        """
        loads configuration from files given in _conf_FILES,
        overwriting the less specific by the most specific (given last)
        """
        for index, filename in enumerate(_CONF_FILES):
            if path.isfile(filename):
                self.read(filename)
            elif index == 0:
                raise CocciException('No package config file: %s' % (filename))


class CocciMatch:
    """
    Store a match and take care of its display
    """

    ptype_regexp = re.compile("^[ )]*\.")

    def __init__(self, mfile, mline, mcol, mlineend, mcolend):
        self.file = mfile
        self.line = int(mline)
        self.column = int(mcol)
        self.lineend = int(mlineend)
        self.columnend = int(mcolend)

    def display(self, stype, mode='raw', oformat='term', before=0, after=0):
        """
        Display output for a single match

        :param mode: display mode
        :type mode: str
        :param oformat: format of output for color (term, html)
        :type oformat: str
        :param before: number of lines to display before match
        :type before: int
        :param after: number of lines to display after match
        :type after: int
        """
        f = open(self.file, 'r')
        lines = f.readlines()
        pmatch = lines[self.line - 1][self.column:self.columnend]
        ptype = "*" # match is a pointer to struct
        if (CocciMatch.ptype_regexp.search(lines[self.line - 1][self.columnend:])):
            ptype = ""
        output = ""
        if mode == 'color':
            output += "%s: l.%s -%d, l.%s +%d, %s %s%s\n" % (self.file,
                 self.line, before, self.line, after, stype, ptype, pmatch)
        for i in range(int(self.line) - 1 - before, int(self.line) + after):
            if mode == 'color':
                output += lines[i]
            elif mode == 'vim':
                output += "%s|%s| (%s %s%s): %s" % (self.file, self.line,
                stype, ptype, pmatch, lines[i])
            elif mode == 'emacs':
                output += "%s:%s: (%s %s%s): %s" % (self.file, self.line,
                stype, ptype, pmatch, lines[i])
            else:
                output += "%s:%s (%s %s%s): %s" % (self.file, self.line,
                stype, ptype, pmatch, lines[i])
        f.close()
        if mode == 'color':
            if have_pygments:
                lexer = CLexer()
                lfilter = NameHighlightFilter(names=[pmatch])
                lexer.add_filter(lfilter)
                if oformat == "term":
                    return highlight(output, lexer, Terminal256Formatter())
                elif oformat == "html":
                    return highlight(output, lexer,
                        HtmlFormatter(noclasses=True))
                else:
                    return output
        return output


class CocciProcess:
    """
    Class used for running spatch command in the case of multiprocessing
    """
    def __init__(self, cmd, verbose):
        self.process = Process(target=self.execute, args=(self, ))
        self.output, self.input = Pipe()
        self.cmd = cmd
        self.verbose = verbose

    def execute(self, option=''):
        output = ""
        try:
            if self.verbose:
                stderr.write("Running: %s.\n" % " ".join(self.cmd))
                output = Popen(self.cmd, stdout=PIPE).communicate()[0]
            else:
                output = Popen(self.cmd, stdout=PIPE,
                    stderr=PIPE).communicate()[0]
        except Exception, err:
            import pickle
            output = pickle.dumps(err)

        self.input.send(output)
        self.input.close()

    def start(self):
        self.process.start()

    def join(self):
        self.process.join()

    def recv(self):
        return self.output.recv()


def _operation_name(fname):
    return path.split(fname)[-1].replace('.cocci', '')


def _raise_run_err(err, cmd):
    if err.errno in (errno.ENOENT, errno.ENOEXEC):
        raise CocciConfigException("Unable to run spatch command "
            "'%s': %s." % (cmd[0], err.strerror))
    raise CocciRunException("Unable to run '%s': %s." % (" ".join(cmd),
        err.strerror))

class CocciPatch:
    """
    Class used to store information about a patch.

    This class is iterable and can be used as a dictionnary.
    """

    keywords = ["Name", "Author", "Desc", "Confidence", "File", "Revision", "Arguments"]
    comment = re.compile("^ *// *(%s): (.*)" % ("|".join(keywords)))

    def __init__(self, filename):
        # open file
        self.__dict__["File"] = filename
        self.__dict__["Name"] = _operation_name(filename)
        # read file
        f = open(filename, 'r')
        lines = f.readlines()
        # match
        for line in lines:
            mm = CocciPatch.comment.match(line)
            if mm:
                self.__dict__[mm.group(1)] = mm.group(2)

    def __iter__(self):
        """ return iterator over keys """
        return self.__dict__.__iter__()

    def __getitem__(self, key):
        # name ok?
        if not key in CocciPatch.keywords:
            raise KeyError("Trying to get invalid name '%s'." % key)
        return self.__dict__[key]

    def __setitem__(self, key, value):
        # name ok?
        if not key in CocciPatch.keywords:
            raise KeyError("Trying to set invalid name '%s'." % key)
        # set
        self.__dict__[key] = value

    def __str__(self):
        out = self["Name"] + ": "
        try:
            out += self["Desc"] + "\n"
        except KeyError:
            return out + ": No info available\n"
        for key in self:
            if key not in ["Name", "Desc", "File"]:
                out +=" * %s: %s\n" % (key, self[key])
        return out

class CocciGrep:
    """
    Core class of the module: setup and run.

    This class is the core of the module. It is responsible
    of initialisation and running of the request.
    """
    spatch = "spatch"
    cocci_python = """

@ script:python @
p1 << init.p1;
@@

for p in p1:
    print "%s:%s:%s:%s:%s" % (p.file,p.line,p.column,p.line_end,p.column_end)
"""

    def __init__(self):
        self.verbose = False
        self.spatch = CocciGrep.spatch
        self.ncpus = 1
        self.operations = {}
        self.process = []
        self.matches = []
        self.options = []
        dirList = listdir(self.get_datadir())
        for fname in dirList:
            op = _operation_name(fname)
            self.operations[op] = path.join(self.get_datadir(), fname)

    def setup(self, stype, attribut, operation):
        """
        :param stype: structure name, used to replace '$type' in the cocci file
        :type stype: str
        :param attribut: basically attribut of the structure, used to replace
            '$attribut' in the cocci file
        :type attribut: str
        :param operation: search operation to do
        :type operation: str
        :raise: :class:`CocciRunException`
        """
        if stype == None:
            raise CocciRunException("Can't use coccigrep without type to "
                "search")
        self.type = stype
        self.attribut = attribut
        self.operation = operation

    def set_concurrency(self, ncpus):
        """
        Set concurrency level (number of spatch command to run in parallel)

        :param ncpus: number of process to launch in parallel
        :type ncpus: int
        """
        if have_multiprocessing:
            self.ncpus = ncpus
            return True
        return False

    def add_options(self, olist):
        """
        Add option to spatch command

        :param olist: List of options
        :type olist: list of str
        """
        self.options += olist

    def set_cpp(self):
        """
        Activate coccinelle C++ support
        """
        self.add_options(["-c++"])

    def set_spatch_cmd(self, cmd):
        """
        Set path or command name for spatch

        :param cmd: Name of parth of the spatch command
        :type cmd: str
        """
        self.spatch = cmd

    def get_datadir(self):
        this_dir, this_filename = path.split(__file__)
        datadir = path.join(this_dir, "data")
        return datadir

    def get_operations(self):
        """
        Get list of available operations

        :return: list of operations in a list of str
        """
        return self.operations.keys()

    def get_operation_name(self, fname):
        return _operation_name(fname)

    def get_operation_info(self, op):
        return CocciPatch(self.operations[op])

    def add_operations(self, new_ops):
        """
        Add operation to the list of supported operations

        :param new_ops: list of filenames (ending by .cocci)
        :type new_ops: list of str
        """
        if len(new_ops) == 0:
            return
        file_filter = re.compile('^[^\.].*\.cocci$')
        for fname in new_ops:
            # file need to end with cocci
            if file_filter.match(fname):
                op = _operation_name(fname)
                self.operations[op] = fname

    def set_verbose(self):
        """
        Activate verbose mode
        """
        self.verbose = True

    def run(self, files):
        """
        Run the search against the files given in argument

        This function is doing the main job. It will run spatch with
        the correct parameters by using subprocess or it will use
        multiprocessing if a concurrency level greater than 1 has been
        asked.

        :param files: list of filenames
        :type files: list of str
        :raise: :class:`CocciRunException` or :class:`CocciConfigException`
        """

        if len(files) == 0:
            raise CocciRunException("Can't use coccigrep without files "
                "to search")
        for cfile in files:
            if not path.isfile(cfile):
                raise CocciRunException("'%s' is not a file, can't "
                    "continue" % cfile)
        # create tmp cocci file:
        tmp_cocci_file = NamedTemporaryFile(suffix=".cocci", delete=False)
        tmp_cocci_file_name = tmp_cocci_file.name
        # open file with name matching operation
        cocci_file = open(self.operations[self.operation], 'r')
        # get the string and build template
        cocci_tmpl = cocci_file.read()
        cocci_smpl_tmpl = Template(cocci_tmpl)
        cocci_file.close()
        cocci_smpl = cocci_smpl_tmpl.substitute(type=self.type,
            attribut=self.attribut)
        cocci_grep = cocci_smpl + CocciGrep.cocci_python

        tmp_cocci_file.write(cocci_grep)
        tmp_cocci_file.close()

        # launch spatch
        output = ""
        # Launch parallel spatch
        if self.ncpus > 1 and len(files) > 1:
            fseq = []
            splitsize = 1.0 / self.ncpus * len(files)
            for i in range(self.ncpus):
                start = int(round(i * splitsize))
                end = int(round((i + 1) * splitsize))
                rfiles = files[start:end]
                if len(rfiles) >= 1:
                    fseq.append(rfiles)
            for sub_files in fseq:
                cmd = [self.spatch]
                cmd += self.options
                cmd += ["-sp_file", tmp_cocci_file.name]
                cmd += sub_files
                sprocess = CocciProcess(cmd, self.verbose)
                sprocess.start()
                self.process.append(sprocess)
            for process in self.process:
                ret = process.recv()
                process.join()
                if not ret.startswith('cexceptions\n'):
                    # CocciProcess return a serialized exception
                    # in case of exception
                    output += ret
                    continue
                import pickle
                err = pickle.loads(ret)
                unlink(tmp_cocci_file_name)
                _raise_run_err(err, cmd)
        # Fallback to one spatch
        else:
            cmd = [self.spatch]
            cmd += self.options
            cmd += ["-sp_file", tmp_cocci_file.name]
            cmd += files
            try:
                if self.verbose:
                    stderr.write("Running: %s.\n" % " ".join(cmd))
                    output = Popen(cmd, stdout=PIPE).communicate()[0]
                else:
                    output = Popen(cmd, stdout=PIPE, stderr=PIPE).communicate()[0]
            except OSError, err:
                unlink(tmp_cocci_file_name)
                _raise_run_err(err, cmd)

            unlink(tmp_cocci_file_name)

        prevfile = None
        prevline = None
        self.matches = []
        for ematch in output.split("\n"):
            try:
                (efile, eline, ecol, elinend, ecolend) = ematch.split(":")
                nmatch = CocciMatch(efile, eline, ecol, elinend, ecolend)
                # if there is equality then we will already display the line
                if (efile == prevfile) and (eline == prevline):
                    continue
                prevfile = efile
                prevline = eline
                self.matches.append(nmatch)
            except ValueError:
                pass

    def display(self, mode='raw', before=0, after=0, oformat='term'):
        """
        Display output for complete request

        :param mode: display mode
        :type mode: str
        :param before: number of lines to display before match
        :type before: int
        :param after: number of lines to display after match
        :type after: int
        :param oformat: format of output for color (term, html)
        :type oformat: str
        :return: the result of the search as a str
        """
        output = ''.join(match.display(self.type, mode=mode, oformat=oformat,
            before=before, after=after)
            for match in self.matches)

        return output.rstrip()
